R/response_process.R

```diff
@@ -0,0 +1,190 @@
+#' Process a Google API response
+#'
+#' @description
+#' Intended primarily for internal use in client packages that provide
+#' high-level wrappers for users. Typically applied as the final step in this
+#' sequence of calls:
+#'   * Request prepared with [request_build()].
+#'   * Request made with [request_make()].
+#'   * Response processed with [response_process()].
+#'
+#' All that's needed for a successful request is to parse the JSON extracted via
+#' `httr::content()`. Therefore, the main point of `response_process()` is to
+#' handle less happy outcomes:
+#'   * Status code in the 100s (information) or 300s (redirection). These are
+#'     unexpected.
+#'   * Non-JSON content type, such as HTML.
+#'   * Status codes in the 400s (client error) and 500s (server error). The
+#'     structure of the error payload varies across Google APIs and we try to
+#'     create a useful message for all variants we know about.
+
+#'
+#' @details
+#' A redacted version of the `resp` input is returned in the condition (auth
+#' tokens are removed). Use functions such as `rlang::last_error()` or
+#' `rlang::catch_cnd()` to capture the condition and do a more detailed forensic
+#' examination.
+#'
+#' @param resp Object of class `response` from [httr].
+#' @param raw Logical. Whether to return parsed content (`FALSE`) or in `raw`
+#'   form.
+#'
+#' @return The content of the request.
+#' @family requests and responses
+#' @export
+response_process <- function(resp, raw = FALSE) {
```

Comment:
Most inspiring ancestor: similar code in bigrquery.

https://github.com/r-dbi/bigrquery/blob/feb1a9ab3019bdd96848acce81cdcfe6770696d6/R/bq-request.R#L164-L206